src/s2i/bin/assemble

```diff
@@ -66,17 +66,20 @@ mv /tmp/src/* "$HOME"
 # set permissions for any installed artifacts
 fix-permissions /opt/app-root -P
 
+{% if spec.version == "2.7" %}
+# Python 2 only code, Python 3 uses bundled wheel from Fedora
 # We have to first upgrade pip to at least 19.3 because:
 # * pip < 9 does not support different packages' versions for Python 2/3
-# * pip < 19.3 does not support manylinux2014 wheels. Only manylinux2014 wheels
+# * pip < 19.3 does not support manylinux2014 wheels. Only manylinux2014 (and later) wheels
 #   support platforms like ppc64le, aarch64 or armv7
-echo "---> Upgrading pip to version 19.3.1 ..."
-if ! pip install -U "pip==19.3.1"; then
-  echo "WARNING: Installation of 'pip==19.3.1' failed, trying again from official PyPI with pip --isolated install"
-  pip install --isolated -U "pip==19.3.1"
+echo "---> Upgrading pip to version <21.0 (The last Python 2 compatible version) ..."
+if ! pip install -U "pip<21.0"; then
+  echo "WARNING: Installation of 'pip<21.0' failed, trying again from official PyPI with pip --isolated install"
+  pip install --isolated -U "pip<21.0"
 fi
+{% endif %}
 
-if [[ ! -z "$UPGRADE_PIP_TO_LATEST"{% if spec.version in ["2.7"] %} || ! -z "$ENABLE_PIPENV"{% endif %} ]]; then
+if [[ ! -z "$UPGRADE_PIP_TO_LATEST" ]]; then
   echo "---> Upgrading pip to latest version ..."
```

Comment:
Let's fix the echo to mention setuptools and wheel as well?

Reply:
Fixed.